src/util/CacheableGenerator.h

```diff
@@ -0,0 +1,288 @@
+                })) {
+          return;
+        }
+        // Timeout exceeded.
+        throw IteratorExpired{};
+      }
+      Timer timer{Timer::Started};
+      if (generatorIterator_.has_value()) {
+        AD_CONTRACT_CHECK(generatorIterator_.value() != generator_.end());
+        ++generatorIterator_.value();
+      } else {
+        generatorIterator_ = generator_.begin();
+      }
+      timer.stop();
+      if (generatorIterator_.value() != generator_.end()) {
+        auto pointer =
+            std::make_shared<T>(std::move(*generatorIterator_.value()));
+        currentSize_.fetch_add(sizeCounter_(*pointer));
+        cachedValues_.push_back(std::move(pointer));
+        if (onSizeChanged_) {
+          onSizeChanged_(std::chrono::milliseconds{timer.msecs()});
+        }
+        tryShrinkCacheIfNeccessary();
+      }
+      if (isMaster) {
+        lock.unlock();
+        conditionVariable_.notify_all();
+      }
+    }
+
+    /// Return the value at `index` or throw an `IteratorExpired` exception if
+    /// it took too long to consume this value.
+    std::shared_ptr<const T> getCachedValue(size_t index) const {
+      std::shared_lock lock{mutex_};
+      if (!cachedValues_.at(index)) {
+        throw IteratorExpired{};
+      }
+      return cachedValues_.at(index);
+    }
+
+    // Needs to be public in order to compile with gcc 11 & 12
+   public:
+    /// Return if the current index is the last of the generator.
+    bool isDone(size_t index) noexcept {
+      std::shared_lock lock{mutex_};
+      return index >= cachedValues_.size() && generatorIterator_.has_value() &&
+             generatorIterator_.value() == generator_.end();
+    }
+
+   private:
+    /// Signal that the master is done processing. After this all of the slaves
+    /// can trigger generation of new entries.
+    void clearMaster() {
+      std::unique_lock lock{mutex_};
+      AD_CORRECTNESS_CHECK(masterState_ != MasterIteratorState::MASTER_DONE);
+      masterState_ = MasterIteratorState::MASTER_DONE;
+      lock.unlock();
+      conditionVariable_.notify_all();
+    }
+
+    /// Set the onSizeChange listener.
+    void setOnSizeChanged(
+        std::function<void(std::optional<std::chrono::milliseconds>)>
+            onSizeChanged) noexcept {
+      std::unique_lock lock{mutex_};
+      onSizeChanged_ = std::move(onSizeChanged);
+    }
+
+    /// Shrink the cache if it exceeds `maxSize_` until either only one element
+    /// is left or we are smaller that `maxSize_` again.
+    void tryShrinkCacheIfNeccessary() {
+      if (currentSize_ <= maxSize_) {
+        return;
+      }
+      size_t maxBound = cachedValues_.size() - 1;
+      for (size_t i = 0; i < maxBound; i++) {
+        auto& pointer = cachedValues_.at(i);
+        if (pointer) {
+          currentSize_.fetch_add(sizeCounter_(*pointer));
+          pointer.reset();
+          if (onSizeChanged_) {
+            onSizeChanged_(std::nullopt);
```

Comment:
This might lead to a deadlock/undefined behaviour by double locking if this leads to deletion of the current onSizeChange listener which needs to acquire this lock again.